File: src/types/telemetry.ts
import type { DbDriver, LogLevel, NormalizedOp, SourceLocation } from "./shared.js";

export interface TelemetryEntry {
  id: string;
  parentRequestId: string | null;
  timestamp: number;
}

export interface TracedFetch extends TelemetryEntry {
  fetchId?: string;
  url: string;
  method: string;
  statusCode: number;
  durationMs: number;
  callSite?: SourceLocation;
}

export interface TracedLog extends TelemetryEntry {
  level: LogLevel;
  message: string;
}

export interface TracedError extends TelemetryEntry {
  name: string;
  message: string;
  stack?: string;
}

export interface TracedQuery extends TelemetryEntry {
  driver: DbDriver;
  sql?: string;
  model?: string;
  operation?: string;
  durationMs: number;
  rowCount?: number;
  normalizedOp?: NormalizedOp;
  table?: string;
  source?: string;
  parentFetchId?: string;
  callSite?: SourceLocation;
}

export type TelemetryEvent =
  | { type: "fetch"; data: Omit<TracedFetch, "id"> }
  | { type: "log"; data: Omit<TracedLog, "id"> }
  | { type: "error"; data: Omit<TracedError, "id"> }
  | { type: "query"; data: Omit<TracedQuery, "id"> };

export type TelemetryBatch = {
  _brakit: true;
  events: TelemetryEvent[];
};

File: sdks/python/brakit/types/telemetry.py
"""Data types for traced telemetry events (queries, fetches, logs, errors)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Union

NormalizedOp = Literal["SELECT", "INSERT", "UPDATE", "DELETE", "OTHER"]

# Python logging levels. Mapped to TS levels by sdk-event-parser.ts LOG_LEVEL_MAP
# (warning -> warn, critical -> error).
LogLevel = Literal["debug", "info", "warning", "error", "critical"]

# Node.js SDK additionally supports: pg, mysql2, prisma
DriverName = Literal["asyncpg", "sqlalchemy", "sdk"]


@dataclass(frozen=True)
class TracedQuery:
    id: str
    parent_request_id: str | None
    timestamp: float
    driver: DriverName
    sql: str | None
    operation: NormalizedOp
    table: str
    duration_ms: float
    row_count: int | None = None
    parent_fetch_id: str | None = None


@dataclass(frozen=True)
class TracedFetch:
    id: str
    parent_request_id: str | None
    timestamp: float
    url: str
    method: str
    status_code: int
    duration_ms: float
    fetch_id: str | None = None


@dataclass(frozen=True)
class TracedLog:
    id: str
    parent_request_id: str | None
    timestamp: float
    level: LogLevel
    message: str


@dataclass(frozen=True)
class TracedError:
    id: str
    parent_request_id: str | None
    timestamp: float
    name: str
    message: str
    stack: str


# Union required at runtime (Python 3.9 compat); X | Y only works in annotations.
TelemetryEntry = Union[TracedQuery, TracedFetch, TracedLog, TracedError]
